feature(js): adds elgg/lightbox AMD module

hypeJunction opened the pull request:
elgg/lightbox AMD module can now be used to open and close the lightbox
programmatically.

Fixes #7895, #8309, #6991

hypeJunction committed:
feature(js): adds elgg/lightbox AMD module

elgg/lightbox AMD module can now be used to open and close the lightbox
programmatically.

Fixes #7895, #8309, #6991

mrclay commented on docs/guides/javascript.rst:
```diff
@@ -542,6 +542,64 @@ Plugins that load a widget layout via Ajax should initialize via this module:
+      'text' => 'Open lightbox',
+      'href' => 'ajax/view/my_view',
+      'class' => 'elgg-lightbox',
+      'data-colorbox-opts' => '{"width: "300px"}'
```
Use json_encode here. A lot more reliable for devs.

mrclay commented on docs/guides/upgrading.rst:
```diff
@@ -17,17 +17,24 @@ Deprecated APIs
 ----------------
 
  * ``elgg/ui.river.js`` is deprecated: Do not rely on simplecache URLs to work.
+ * ``lightbox/settings.js`` is deprecated: Use ``getOptions,ui.lightbox`` plugin hook or ``data-colorbox-opts`` attribute.
```
Clarify this is a JS hook (register in a boot plugin)

mrclay commented on views/default/elgg.js.php:
```diff
@@ -110,4 +110,16 @@
+require(['elgg/init', 'elgg/ready']);
+
+<?php
+if (_elgg_view_may_be_altered('lightbox/settings.js', '/lightbox/settings.js.php')) {
```
Don't include leading / in this path. We may have to use that to distinguish absolute file paths.

mrclay commented on views/default/elgg/lightbox.js:
```diff
@@ -0,0 +1,98 @@
+	var elgg = require('elgg');
+	var $ = require('jquery');
+	require('jquery.colorbox');
+	
```
Consumes a plugin hook so it should depend on elgg/init as well.

mrclay commented:
This is a heroic effort that looks really close.

mrclay commented on docs/guides/javascript.rst:
```diff
@@ -542,6 +542,64 @@ Plugins that load a widget layout via Ajax should initialize via this module:
    });
 
+Module ``elgg/lightbox``
+-----------------------
```
One more hyphen

hypeJunction commented:
@mrclay Addressed your comments
@Elgg/contributors @Elgg/core Anyone opposed to loading lightbox by default? In the end, it seems that most sites end up loading it on each page request anyway, e.g. if you enable reportedcontent.

hypeJunction commented:
One other option is to extend `elgg.css` with the colorbox theme, and then use event binding in `elgg.ui.init()` as so:

```js
$(document).on('click', '.elgg-lightbox', function(e) {
   e.preventDefault();
   var dataOpts = $(this).data('colorboxOpts');
   require(['elgg/lightbox'], function(lightbox) {
      lightbox.open(lightbox.getOptions(dataOpts);
   });
});
```

hypeJunction commented on docs/guides/javascript.rst:
```diff
@@ -542,6 +542,65 @@ Plugins that load a widget layout via Ajax should initialize via this module:
+
+   require(['elgg/lightbox'], function() {
+      $('.my-gallery').colorbox({photo: true});
+   });
```
Any idea what Travis wants from me?
```sh
Warning, treated as error:
/home/travis/build/Elgg/Elgg/docs/guides/javascript.rst:602: WARNING: Definition list ends without a blank line; unexpected unindent.
```

jeabakker commented:
> Anyone opposed to loading lightbox by default? In the end, it seems that most sites end up loading it on each page request anyway, e.g. if you enable reportedcontent.

on principle i don't like loading stuff you're not gonna use. If this is CSS i can sort of live with it, but for JS it has more impact on the UX so if it can be avoided i'm all for it

hypeJunction commented:
I don't see how it affects UX other than performance, which will be negligible with simplecache enabled. It does however make developer's life easier. Until we have a way to load css on demand with js, this would be my preferred approach. I think we often make bad UX decisions due to additional hassle it takes to load/configure/position components. I am doing my best to rectify that.

mrclay commented:
Putting the click binding in elgg.ui.init() feels like mixing concerns. What if elgg/lightbox just had this binding code, but it async required elgg/lightbox/api, which had the getOptions/open/close. Then we could inline elgg/lightbox in elgg.js

hypeJunction commented:
I don't see how that helps our cause.

mrclay commented on docs/guides/javascript.rst:
```diff
@@ -542,6 +542,65 @@ Plugins that load a widget layout via Ajax should initialize via this module:
+
+To support gallery sets (via ``rel`` attribute), you need to bind colorbox to a specific selector:
+
+..code:: js
```
`.. code:: js`

mrclay commented on docs/guides/javascript.rst:
```diff
@@ -542,6 +542,65 @@ Plugins that load a widget layout via Ajax should initialize via this module:
+..code:: js
+
+   require(['elgg/lightbox'], function() {
+      $('.my-gallery').colorbox({photo: true});
```
We should add a lightbox "bind" method so we don't instruct users to call colorbox API directly.

hypeJunction replied:
We have a `bind` method but it does something else.

hypeJunction commented:
I like the idea of `data-colorbox-opts` but it seems it to complicate things. We could try to use `onOpen` method to populate options with element specific values.

hypeJunction commented:
Colorbox natively supports `data-cbox-` attributes. We can deprecate `data-colorbox-opts` and use simple `$(selector).colorbox(opts)` bindings, which will support galleries out of the box.

mrclay commented:
The point is we shouldn't put some lightbox code into the UI library just for performance. Better to keep lightbox code in AMD module(s) and inline those as needed for performance. Maybe clearer: we move the event binding to a elgg/lightbox/init module and just inline that.

`data-cbox` is undocumented; we should consider it private API.

hypeJunction commented:
This PR does what it is meant for. I am fine getting this in and improving as we move along.

mrclay commented:
Agreed. LGTM

hypeJunction commented:
Slight modifications:
- Now also supports `.elgg-lightbox-iframe` and `.elgg-lightbox-inline`
- Clears click bindings before applying new ones to avoid duplicate bindings
- `elgg/lightbox#bind` now takes 3rd attribute to allow binding directly via `$(selector).colorbox()` to make it easier to support groups of related elements
- Added examples to theme sandbox

mrclay commented on views/default/lightbox.js.php:
```diff
@@ -30,88 +23,89 @@
+		 */
+		elgg.ui.lightbox.getSettings = function () {
+			elgg.deprecated_notice('elgg.ui.lightbox.getSettings() has been deprecated and should not be called directly. Use elgg/lightbox AMD module intsead', '2.2');
+			var lightbox = require('elgg/lightbox');
```
Are we inlining elgg/lightbox?

hypeJunction replied:
Not in this PR.

hypeJunction replied:
Any other ideas on how to make this work sync?

mrclay replied:
I'd be OK inlining it to get this done for 2.x.

hypeJunction replied:
Hm, we depend on `jquery.colorbox`

mrclay replied:
Have you tested that sync `require('module')` actually works?

hypeJunction replied:
It did, but I suppose because we load elgg/lightbox in init
On Apr 5, 2016 4:26 PM, "Steve Clay" <notifications@github.com> wrote:

> In views/default/lightbox.js.php
> <https://github.com/Elgg/Elgg/pull/9611#discussion_r58547112>:
>
> > -};
> > -
> > -/**
> > - * Close the colorbox
> > - *
> > - */
> > -elgg.ui.lightbox.close = function() {
> > -	$.colorbox.close();
> > -};
> > +		/**
> > +		 * Lightbox settings
> > +		 * @deprecated 2.2
> > +		 */
> > +		elgg.ui.lightbox.getSettings = function () {
> > +			elgg.deprecated_notice('elgg.ui.lightbox.getSettings() has been deprecated and should not be called directly. Use elgg/lightbox AMD module intsead', '2.2');
> > +			var lightbox = require('elgg/lightbox');
>
> Have you tested that sync require('module') actually works?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/9611/files/c3a290dc30d2aa697777932e6f0750bb5615044d#r58547112>
>

mrclay commented:
Considering sync require() isn't documented (and we likely aren't testing it with a range of browsers), I think it's unwise to rely on it. What if we just left the code in elgg.ui.lightbox as is?

hypeJunction commented:
I am working on a solution. Coming shortly.

hypeJunction committed:
feature(js): lightbox is now bootstrapped at runtime

Lightbox is now bootstrapped at runtime. Colorbox however is not required until
elgg/lightbox#open is called. Lightbox css now extends elgg.css.

hypeJunction commented:
@mrclay Thoughts? If we like this, we will need to remove calls to ``elgg_load_js`` and ``elgg_load_css`` in core.

hypeJunction commented on views/default/elgg.js.php:
```diff
@@ -25,9 +23,11 @@
+// as these modules will be required on each page
 echo elgg_view('elgg/popup.js');
+echo elgg_view('elgg/lightbox.js');
+echo elgg_view('elgg/ui/lightbox.js');
```
Could also be ``elgg/init/lightbox``, we could throw all init code in there.

mrclay commented:
Just copy the elgg/lightbox#getOptions code back into `elgg.ui.lightbox.getSettings`. That shouldn't introduce any bugs and keeps it synchronous. A little duplication isn't going to kill us.

With that change would we have to inline anything? Would there still be reason to keep a separate module to do the binds?

mrclay commented:
Mind if I send a PR?

hypeJunction commented:
Feel free. I am going to retire for a few days to focus on other things.
On Apr 5, 2016 6:01 PM, "Steve Clay" <notifications@github.com> wrote:

> Mind if I send a PR?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Elgg/Elgg/pull/9611#issuecomment-205870352>
>

mrclay commented:
Consider #9625

hypeJunction commented:
Replaced by #9625
